torch/csrc/autograd/engine.cpp

```diff
@@ -271,6 +280,10 @@ auto Engine::thread_main(GraphTask *graph_task) -> void {
   // Note [Reentrant backwards]
   while (!graph_task || graph_task->outstanding_tasks_ > 0) {
     FunctionTask task = queue->pop();
+    if (task.isShutdownTask_) {
+      C10_LOG_API_USAGE_ONCE("worker shutting down");
+      break;
```

Comment:
When using  reentrant autograd, the `thread_main` function is called again, with a specific graph_task. So the shutdown task could be caught by a `thread_main` from the reentrant autograd and not the original one that keeps the thread alive.
Or do you assume that no backward us currently running when the engine is destroyed? If so, this should be checked in some way.

Reply:
We expect this to work only when no backward is running while the engine is destroyed. This might not always be the case though and it needs a separate fix. Could you clarify when needs to be checked?

Reply:
I guess making sure all ready queues are empty will make sure that no backward is currently running.

Reply:
My previous comment wasn't entirely correct. I think it's okay for a backward to be running if it's not from a reentrant autograd, so the queue may not be empty. Should we only process the shutdown task in those cases?

Reply:
What should be the behavior if the Engine is destroyed while a backward is running? Wait for it to finish? Or interrupt it?
I guess that will change the priority of the task your add, minimal if you want to wait for it to finish and maximal otherwise.
In both cases, I would says that another shutdown task should be queued if !graph_task: inside reentrant autograd, (optionally stop the current processing) and notify the other thread_main that it should exit.

Reply:
We want to interrupt it, so the priority is currently maximal. 
I had discussed re-queuing a shutdown task from inside a reentrant autograd with @ezyang  offline and we decided to go with this approach for now, but we can still consider it. @ezyang Could you comment on this?

Reply:
@albanD My hope was to pretend this problem didn't exist :)

I think the correct thing to do is to stop executing work as much as soon as possible, throwing exceptions if necessary to speed things up. We just have to be careful to catch these exceptions properly and then swallow them (don't report errors to users here). If you queue another task on the queue, you'll wait for the task which made the reentrant backwards call to finish executing the rest of its stuff before shutting down; if we are actually going to fix this edge case, let's fix it properly :)

Reply:
In that case what happens is that the "parent" `thread_main` is stuck in its `evaluate_function`.
The "child" `thread_main` is executing stuff.
So in case of reentrant, the "child" will always get the shutdown task first.
That will interrupt it and force it to exit early.
Now the "parent" exit its `evaluate_function` and will start running other stuff.
If another shutdown task was not queued, it will try to finish it's backward and the thread will never be destroyed.

Maybe preventing destruction when a backward is running is too much? Or could be done as future work if it is actually needed.
Checking that all queues are empty before queuing the shutdown task seems ok for me. And raise an error saying that the Engine cannot be destroyed if it's running stuff. That would be simpler.
@ezyang does that sound ok with you?

Reply:
Yes, SGTM. Actually I wouldn't raise an error, because I am not sure if we run destructors when you C-C a PyTorch process (I don't know exactly what Python's signal handler does...)

Reply:
I'm pretty sure Python catches C-C, turns it into a `KeyboardException` and then terminates the program "cleanly" (in the sense that the destructors will always run). I have no idea which thread will be responsible for the cleanup though...

Reply:
It wouldn't surprise me if all of this handling happened on the main thread.